File: 0207-course-schedule/0207-course-schedule.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
var canFinish = function(numCourses, prerequisites) {
    const graph = buildGraph(numCourses, prerequisites)  
    const visited = new Set()
    const visiting = new Set()

    for (const course in graph) {
      if (hasCycle(graph, course, visited, visiting)) {
          return false
      }
    }
    
    return true
};

const hasCycle = (graph, course, visited, visiting) => {
    if (visited.has(course)) return false
    if (visiting.has(course)) return true
    
    visiting.add(course)
    
    for (const neighbor of graph[course]) {
        if (hasCycle(graph, neighbor, visited, visiting)) {
            return true
        }
    }
    
    visiting.delete(course)
    visited.add(course)
    
    return false
}

const buildGraph = (numCourses, prereqs) => {
    const graph = {}
    
    for (let i = 0; i < numCourses; i++) {
        graph[i] = []
    }
    
    for (const [a, b] of prereqs) {
        graph[a].push(b)
    }
    
    return graph;
}

File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = self.build_graph(numCourses, prerequisites)
        visited = set()
        visiting = set()
        
        for course in graph:
            if self.has_cycle(graph, course, visited, visiting) == True:
                return False
            
        return True
    
    def has_cycle(self, graph: dict, course: int, visited: set, visiting: set) -> bool:
        if course in visited:
            return False
        if course in visiting:
            return True
        
        visiting.add(course)
        
        for neighbor in graph[course]:
            if self.has_cycle(graph, neighbor, visited, visiting) == True:
                return True
            
        visiting.remove(course)
        visited.add(course)
        
        return False
        
        
    def build_graph(self, numCourses: int, prerequisites: List[List[int]]) -> dict:
        graph = { i:[] for i in range(numCourses) }
        
        for a, b in prerequisites:
            graph[a].append(b)
            
        return graph
        